pkg/helm/watcher/controller/generation_update_predicate.go

```diff
@@ -0,0 +1,20 @@
+package controller
+
+import (
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// GenerationUpdatePredicate triggers an update event when a HelmRepository generation changes.
+// i.e.: Delete events.
+type GenerationUpdatePredicate struct {
+	predicate.Funcs
+}
+
+func (GenerationUpdatePredicate) Update(e event.UpdateEvent) bool {
+	if e.ObjectOld == nil || e.ObjectNew == nil {
+		return false
+	}
+
+	return e.ObjectOld.GetGeneration() < e.ObjectNew.GetGeneration()
```

Comment:
when would this not be the case :thinking:

Reply:
Yes.

Because...

How this works is as follows. The HelmRepository is created. At that point, it's generation 1. This is the create event, we don't care about that since it doesn't have an artifact. Then flux reconciles it, making it generation two, but artifact is still not updated. Then comes a patch event the artifact is added, but the generation doesn't change. To illustrate: 

This is an update event where the artifact is added. Called with old and called with new are the two artifact that come in. Here, the observed generation stays two, but for the first artifact is `nil` for the second, artifact is there.
```
called with old: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:podinfo GenerateName: Namespace:wego-system SelfLink: UID:78d8620a-f19c-4fa0-a8e3-33c62775d050 ResourceVersion:41538 Generation:2 CreationTimestamp:2022-01-14 17:20:51 +0100 CET DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[kustomize.toolkit.fluxcd.io/name:kind-kind-user kustomize.toolkit.fluxcd.io/namespace:wego-system] Annotations:map[kubectl.kubernetes.io/last-applied-configuration:{"apiVersion":"source.toolkit.fluxcd.io/v1beta1","kind":"HelmRepository","metadata":{"annotations":{"kustomize.toolkit.fluxcd.io/checksum":"57c8b5346547f7eba33632994f4b57dd5cf744db"},"labels":{"kustomize.toolkit.fluxcd.io/name":"kind-kind-user","kustomize.toolkit.fluxcd.io/namespace":"wego-system"},"name":"podinfo","namespace":"wego-system"},"spec":{"interval":"30s","url":"https://weaveworks.github.io/profiles-examples"}}
 kustomize.toolkit.fluxcd.io/checksum:57c8b5346547f7eba33632994f4b57dd5cf744db] OwnerReferences:[] Finalizers:[finalizers.fluxcd.io finalizers.helm.watcher] ClusterName: ManagedFields:[{Manager:gitops Operation:Update APIVersion:source.toolkit.fluxcd.io/v1beta1 Time:2022-01-14 17:20:51 +0100 CET FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:finalizers":{"v:\"finalizers.helm.watcher\"":{}}}} Subresource:} {Manager:kustomize-controller Operation:Update APIVersion:source.toolkit.fluxcd.io/v1beta1 Time:2022-01-14 17:20:51 +0100 CET FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:annotations":{".":{},"f:kubectl.kubernetes.io/last-applied-configuration":{},"f:kustomize.toolkit.fluxcd.io/checksum":{}},"f:labels":{".":{},"f:kustomize.toolkit.fluxcd.io/name":{},"f:kustomize.toolkit.fluxcd.io/namespace":{}}},"f:spec":{".":{},"f:interval":{},"f:url":{}}} Subresource:} {Manager:source-controller Operation:Update APIVersion:source.toolkit.fluxcd.io/v1beta1 Time:2022-01-14 17:20:51 +0100 CET FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:finalizers":{".":{},"v:\"finalizers.fluxcd.io\"":{}}},"f:spec":{"f:timeout":{}},"f:status":{".":{},"f:conditions":{},"f:observedGeneration":{}}} Subresource:}]} Spec:{URL:https://weaveworks.github.io/profiles-examples SecretRef:<nil> PassCredentials:false Interval:{Duration:30s} Timeout:&Duration{Duration:1m0s,} Suspend:false AccessFrom:<nil>} Status:{HERE =>>>>>>>>>>ObservedGeneration:2 Conditions:[{Type:Ready Status:Unknown ObservedGeneration:0 LastTransitionTime:2022-01-14 17:20:51 +0100 CET Reason:Progressing Message:reconciliation in progress}] URL: HERE >>>>>>>>>>>Artifact:<nil> ReconcileRequestStatus:{LastHandledReconcileAt:}}}



called with new: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:podinfo GenerateName: Namespace:wego-system SelfLink: UID:78d8620a-f19c-4fa0-a8e3-33c62775d050 ResourceVersion:41540 Generation:2 CreationTimestamp:2022-01-14 17:20:51 +0100 CET DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[kustomize.toolkit.fluxcd.io/name:kind-kind-user kustomize.toolkit.fluxcd.io/namespace:wego-system] Annotations:map[kubectl.kubernetes.io/last-applied-configuration:{"apiVersion":"source.toolkit.fluxcd.io/v1beta1","kind":"HelmRepository","metadata":{"annotations":{"kustomize.toolkit.fluxcd.io/checksum":"57c8b5346547f7eba33632994f4b57dd5cf744db"},"labels":{"kustomize.toolkit.fluxcd.io/name":"kind-kind-user","kustomize.toolkit.fluxcd.io/namespace":"wego-system"},"name":"podinfo","namespace":"wego-system"},"spec":{"interval":"30s","url":"https://weaveworks.github.io/profiles-examples"}}
 kustomize.toolkit.fluxcd.io/checksum:57c8b5346547f7eba33632994f4b57dd5cf744db] OwnerReferences:[] Finalizers:[finalizers.fluxcd.io finalizers.helm.watcher] ClusterName: ManagedFields:[{Manager:gitops Operation:Update APIVersion:source.toolkit.fluxcd.io/v1beta1 Time:2022-01-14 17:20:51 +0100 CET FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:finalizers":{"v:\"finalizers.helm.watcher\"":{}}}} Subresource:} {Manager:kustomize-controller Operation:Update APIVersion:source.toolkit.fluxcd.io/v1beta1 Time:2022-01-14 17:20:51 +0100 CET FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:annotations":{".":{},"f:kubectl.kubernetes.io/last-applied-configuration":{},"f:kustomize.toolkit.fluxcd.io/checksum":{}},"f:labels":{".":{},"f:kustomize.toolkit.fluxcd.io/name":{},"f:kustomize.toolkit.fluxcd.io/namespace":{}}},"f:spec":{".":{},"f:interval":{},"f:url":{}}} Subresource:} {Manager:source-controller Operation:Update APIVersion:source.toolkit.fluxcd.io/v1beta1 Time:2022-01-14 17:20:51 +0100 CET FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:finalizers":{".":{},"v:\"finalizers.fluxcd.io\"":{}}},"f:spec":{"f:timeout":{}},"f:status":{".":{},"f:artifact":{".":{},"f:checksum":{},"f:lastUpdateTime":{},"f:path":{},"f:revision":{},"f:url":{}},"f:conditions":{},"f:observedGeneration":{},"f:url":{}}} Subresource:}]} Spec:{URL:https://weaveworks.github.io/profiles-examples SecretRef:<nil> PassCredentials:false Interval:{Duration:30s} Timeout:&Duration{Duration:1m0s,} Suspend:false AccessFrom:<nil>} Status:{HERE =>>>>>>>>> ObservedGeneration:2 Conditions:[{Type:Ready Status:True ObservedGeneration:0 LastTransitionTime:2022-01-14 17:20:51 +0100 CET Reason:IndexationSucceed Message:Fetched revision: eca4147a15d2ac4a5eeca1cf247ac25505ae831c}] URL:http://source-controller.wego-system.svc.cluster.local./helmrepository/wego-system/podinfo/index.yaml HERE =>>>>>>>>>>>>>>>Artifact:0xc0048d1440 ReconcileRequestStatus:{LastHandledReconcileAt:}}}
```

You can see that generation stays, but artifact has been changed. So we need two predicates. One observing artifacts which are update events. And one observing the generation so we also catch and deal with Delete events.

Reply:
Ok, discussion outcome:

- First, I was wrong. :) The req contains the name and namespace which would be enough to delete the cache. No idea why I didn't spot that. :D
- Second, we still opted to keep the finaliser because we want to make sure that a resource only gets deleted once we successfully deleted the cache as well, otherwise we would be showing stale data on the UI. Still saying profile exists even though the helmrepository was deleted
- Third, there is an optimization possible later where we also check the revision of the new Generation of the resource so we don't update the cache if only the interval is changed for example. ( Interval and spec change increase the generation of the object ) ( I'll create a ticket about that to further improve the inner working of the reconcile loop )